### DISCUSSION TITLE
Issue with MariaDB "Got error: 2002: "Can't connect to local server through socket '/tmp/mysql.sock' (2)" when trying to connect"

### DISCUSSION THREAD

**Discussion by YousufSSyed** [Homebrew/discussions#3410] [Everyday usage]
I'm unable to startup Mariadb, spent a while troubleshooting, and haven't found any solutions. At this point, I just want it working so that I can export my database and switch to running Mariadb on [Lima](https://github.com/lima-vm/lima), because Mariadb on Homebrew is simply that unreliable. I ran all the commands outputted by `brew doctor`, ran `brew reinstall mariadb`, and a few other things a while ago that I forgot, but I still get the same error. For each things I tried, I ran `brew services restart mariadb`. I'm on macOS 12.4.

I'm going to be very patient with fixing if I have to do lots of back and forth with troubleshooting.

**Edit:** I'm using an M1 Mac.

**Comment by SMillerDev** [comment]
What does `brew services` say about mariadb? Is there actually a `/tmp/mysql.sock` file to use?

**Reply by YousufSSyed** [reply]
`brew services` outputs this for mariadb:

`mariadb    started root ~/Library/LaunchAgents/homebrew.mxcl.mariadb.plist`

**Reply by YousufSSyed** [reply]
`/tmp/mysql.sock` doesn't exist after running `brew services start mariadb` (and that doesn't output any errors)

**Reply by SMillerDev** [reply]
Can you try stopping the service using `brew services stop mariadb` and running `/usr/local/opt/mariadb/bin/mysqld_safe --datadir=/usr/local/var/mysql` instead and sending the output here.

**Reply by YousufSSyed** [reply]
I think that's the directory for Intel Macs. I should mention I'm using an M1 Mac. it says on [brew.sh](https://docs.brew.sh/Installation): This script installs Homebrew to its preferred prefix (`/usr/local` for macOS Intel, `/opt/homebrew` for Apple Silicon 

the mysql folder is the folder with the one the contains folders for each database and files like `ib_logfile0`, correct? I think this would the right directory: `/opt/homebrew/var/mysql`. So I tried the command `/opt/homebrew/opt/mariadb/bin/mysqld_safe --datadir=/opt/homebrew/var/mysql`, and this is what I got:

```
/opt/homebrew/opt/mariadb/bin/my_print_defaults: Can't read dir of '/opt/homebrew/etc/my.cnf.' (Errcode: 2 "No such file or directory")
Fatal error in defaults handling. Program aborted
/opt/homebrew/opt/mariadb/bin/my_print_defaults: Can't read dir of '/opt/homebrew/etc/my.cnf.' (Errcode: 2 "No such file or directory")
Fatal error in defaults handling. Program aborted
220706 06:22:42 mysqld_safe Logging to '/opt/homebrew/var/mysql/MacBook-Pro.local.err'.
220706 06:22:42 mysqld_safe Starting mariadbd daemon with databases from /opt/homebrew/var/mysql
```
But I can see that the file `/opt/homebrew/etc/my.cnf` exists in Finder. You might want to note the distinction between `my.cnf` and `my.cnf.` shown in the output.

**Reply by SMillerDev** [reply]
What does `ls -al /opt/homebrew/etc/my.cnf*` say?

**Reply by YousufSSyed** [reply]
```
-rw-r--r--@ 1 yousuf  admin  204 Jul  6 04:28 /opt/homebrew/etc/my.cnf
-rwxrwxrwx  1 yousuf  admin  206 Jul  6 06:17 /opt/homebrew/etc/my.cnf.default
-rwxrwxrwx  1 yousuf  admin  113 Jul  6 06:17 /opt/homebrew/etc/my.cnf.default.default

/opt/homebrew/etc/my.cnf.d:
total 0
drwxrwxrwx   3 yousuf  admin    96 Apr 10 09:02 .
drwxrwxrwx  35 yousuf  admin  1120 Jul  6 04:27 ..
-rwxrwxrwx   1 yousuf  admin     0 Jul  6 06:17 .homebrew_dont_prune_me
```

**Reply by SMillerDev** [reply]
And what is the content of `/opt/homebrew/etc/my.cnf`?

**Reply by carlocab** [reply]
> `Can't read dir of '/opt/homebrew/etc/my.cnf.'`

That extra `.` looks suspicious

**Reply by gromgit** [reply]
>```
> -rw-r--r--@ 1 yousuf  admin  204 Jul  6 04:28 /opt/homebrew/etc/my.cnf
>```

The `@` sign after the permissions says this file has extended attributes, one or more of which may be preventing you from opening it.

What's the output of `ls -ld@ /opt/homebrew/etc/my.cnf`?

**Comment by YousufSSyed** [comment]
@carlocab Yeah I thought the . might be noteworthy.

@gromgit `ls -ld@ /opt/homebrew/etc/my.cnf` outputs this:
```
-rw-r--r--@ 1 yousuf  admin  204 Jul  6 04:28 /opt/homebrew/etc/my.cnf
	com.apple.LaunchServices.OpenWith	150 
	com.apple.lastuseddate#PS	 16
```

(I deleted the comment since the section was getting overflown).

**Reply by gromgit** [reply]
That doesn't look like a deal-breaker.

Returning to your original error, what's the output of `/opt/homebrew/opt/mariadb/bin/my_print_defaults`?

**Reply by YousufSSyed** [reply]
Just so you know, I'm _not_ an expert with the SQL language or in database administration. This doesn't look like a proper command looking at the output, but here it is when I enter `/opt/homebrew/opt/mariadb/bin/my_print_defaults` in the terminal:

```
/opt/homebrew/opt/mariadb/bin/my_print_defaults  Ver 1.7 for osx10.17 at arm64
This software comes with ABSOLUTELY NO WARRANTY. This is free software,
and you are welcome to modify and redistribute it under the GPL license

Displays the options from option groups of option files, which is useful to see which options a particular tool will use
Usage: /opt/homebrew/opt/mariadb/bin/my_print_defaults [OPTIONS] [groups]
  -#, --debug[=#]     This is a non-debug version. Catch this and exit
  --mysqld            Read the same set of groups that the mysqld binary does.
  -n, --no-defaults   Return an empty string (useful for scripts).
  -?, --help          Display this help message and exit.
  -v, --verbose       Increase the output level
  -V, --version       Output version information and exit.

Default options are read from the following files in the given order:
/opt/homebrew/etc/my.cnf ~/.my.cnf 

Variables (--variable-name=value)
and boolean options {FALSE|TRUE}  Value (after reading options)
--------------------------------- ----------------------------------------
mysqld                            FALSE

Example usage:
/opt/homebrew/opt/mariadb/bin/my_print_defaults --defaults-file=example.cnf client client-server mysql
```
And when I enter `/opt/homebrew/opt/mariadb/bin/my_print_defaults --defaults-file=example.cnf client client-server mysql`: 

```
defaults-file=example.cnf client client-server mysql
Could not open required defaults file: /opt/homebrew/example.cnf
Fatal error in defaults handling. Program aborted
```

**Reply by SMillerDev** [reply]
And what is the content of `/opt/homebrew/etc/my.cnf`?

**Reply by YousufSSyed** [reply]
```
#
# This group is read both by the client and the server
# use it for options that affect everything
#
[client-server]

#
# include *.cnf from the config directory
#
!includedir /opt/homebrew/etc/my.cnf.d
```

**Reply by gromgit** [reply]
> This doesn't look like a proper command looking at the output

It is a proper command, I was really looking for this output:
>```
>Default options are read from the following files in the given order:
>/opt/homebrew/etc/my.cnf ~/.my.cnf 
>```

to verify that the command wasn't looking in some weird place like, well, `/opt/homebrew/etc/my.cnf.` (with the trailing period).

Is there a `~/.my.cnf` on your system? If so, what is its contents?

**Reply by YousufSSyed** [reply]
`/opt/homebrew/etc/my.cnf.` outputs `No such file or directory`. Same thing with `~/.my.cnf`

**Reply by SMillerDev** [reply]
How about a `/etc/my.cnf`?

**Reply by YousufSSyed** [reply]
It's either an alias or symlink to `/opt/homebrew/etc/my.cnf` (I'm pretty sure it's a symlink).

**Reply by YousufSSyed** [reply]
Ok I just figured out it's a symlink.

**Comment by gromgit** [comment]
What are the contents of `/opt/homebrew/var/mysql/MacBook-Pro.local.err`?

**Reply by YousufSSyed** [reply]
Here is the file. It's so big, I had to compress it! 58.5MB uncompressed!

[MacBook-Pro.local.err.txt.zip](https://github.com/Homebrew/discussions/files/9064081/MacBook-Pro.local.err.txt.zip)

**Reply by gromgit** [reply]
Not a MariaDB user, but the log is pretty informative (it's so large because MariaDB is continually failing to restart):
```
220706 03:43:45 mysqld_safe Starting mariadbd daemon with databases from /opt/homebrew/var/mysql
2022-07-06  3:43:45 0 [Note] /opt/homebrew/opt/mariadb/bin/mariadbd (server 10.8.3-MariaDB) starting as process 26485 ...
2022-07-06  3:43:45 0 [Warning] Can't create test file /opt/homebrew/var/mysql/MacBook-Pro.lower-test
2022-07-06  3:43:45 0 [ERROR] mariadbd: Can't create/write to file './ddl_recovery.log' (Errcode: 13 "Permission denied")
2022-07-06  3:43:45 0 [ERROR] DDL_LOG: Failed to create ddl log file: ./ddl_recovery.log
2022-07-06  3:43:45 0 [ERROR] Aborting
2022-07-06  3:43:45 0 [Note] /opt/homebrew/opt/mariadb/bin/mariadbd (server 10.8.3-MariaDB) starting as process 26435 ...
2022-07-06  3:43:45 0 [Warning] Setting lower_case_table_names=2 because file system for /opt/homebrew/var/mysql/ is case insensitive
/opt/homebrew/opt/mariadb/bin/mariadbd: Please consult the Knowledge Base to find out how to run mysqld as root!
2022-07-06  3:43:45 0 [ERROR] Aborting
220706 03:43:45 mysqld_safe mysqld from pid file /opt/homebrew/var/mysql/MacBook-Pro.local.pid ended
```
Failure to create any log file is not a good sign, and their knowledge base is quite clear that [MariaDB should not be run as root](https://mariadb.com/kb/en/running-mysqld-as-root/).
```
220706 06:20:27 mysqld_safe Starting mariadbd daemon with databases from /opt/homebrew/var/mysql
Warning: World-writable config file '/opt/homebrew/etc/my.cnf' is ignored
2022-07-06  6:20:27 0 [Note] /opt/homebrew/opt/mariadb/bin/mariadbd (server 10.8.3-MariaDB) starting as process 61844 ...
2022-07-06  6:20:27 0 [Warning] Setting lower_case_table_names=2 because file system for /opt/homebrew/var/mysql/ is case insensitive
2022-07-06  6:20:27 0 [Note] InnoDB: Compressed tables use zlib 1.2.11
2022-07-06  6:20:27 0 [Note] InnoDB: Number of transaction pools: 1
2022-07-06  6:20:27 0 [Note] InnoDB: Using generic crc32 instructions
2022-07-06  6:20:27 0 [Note] InnoDB: Initializing buffer pool, total size = 128.000MiB, chunk size = 2.000MiB
2022-07-06  6:20:27 0 [Note] InnoDB: Completed initialization of buffer pool
2022-07-06  6:20:27 0 [ERROR] InnoDB: Upgrade after a crash is not supported. The redo log was created with MariaDB 10.6.4.
2022-07-06  6:20:27 0 [ERROR] InnoDB: Plugin initialization aborted with error Generic error
2022-07-06  6:20:27 0 [Note] InnoDB: Starting shutdown...
2022-07-06  6:20:27 0 [ERROR] Plugin 'InnoDB' init function returned error.
2022-07-06  6:20:27 0 [ERROR] Plugin 'InnoDB' registration as a STORAGE ENGINE failed.
2022-07-06  6:20:27 0 [Note] Plugin 'FEEDBACK' is disabled.
2022-07-06  6:20:27 0 [ERROR] Unknown/unsupported storage engine: InnoDB
2022-07-06  6:20:27 0 [ERROR] Aborting
220706 06:20:27 mysqld_safe mysqld from pid file /opt/homebrew/var/mysql/MacBook-Pro.local.pid ended
```
Note the world-writable `my.cf` warning. The `Upgrade after a crash` error seems fatal, `Plugin 'InnoDB' registration as a STORAGE ENGINE failed` definitely so. Whatever data you may have had in the DB should be considered lost for good, and your current installation is probably not salvageable. I suggest the following:
* `sudo brew services stop mariadb`
* `sudo brew services cleanup`
* `sudo pkill -KILL -f mysql mariadb` to ensure every process with `mysql` or `mariadb` in its name dies a horrible death (NOTE: **NEVER USE** `pkill -KILL` under normal circumstances)
* `brew remove mariadb`
* clean up all the remaining MariaDB artifacts, especially `/opt/homebrew/etc/my.cf` and the entire `/opt/homebrew/var/mysql` directory
* `brew install mariadb`
* `/opt/homebrew/opt/mariadb/bin/mysqld_safe --datadir=/opt/homebrew/var/mysql` to ensure it can be started without error
* `brew services start mariadb` to put it under launchd control

**Reply by maazmmd** [reply]
@gromgit @YousufSSyed  I'm having similar issue, been stuck for a while. Here's content from log file

```sh
240716 21:16:42 mysqld_safe Starting mariadbd daemon with databases from /opt/homebrew/var/mysql
2024-07-16 21:16:42 0 [Warning] Setting lower_case_table_names=2 because file system for /opt/homebrew/var/mysql/ is case insensitive
2024-07-16 21:16:42 0 [Note] Starting MariaDB 11.4.2-MariaDB source revision 3fca5ed772fb75e3e57c507edef2985f8eba5b12 as process 33699
2024-07-16 21:16:42 0 [Note] InnoDB: Compressed tables use zlib 1.2.12
2024-07-16 21:16:42 0 [Note] InnoDB: Number of transaction pools: 1
2024-07-16 21:16:42 0 [Note] InnoDB: Using generic crc32 instructions
2024-07-16 21:16:42 0 [Note] InnoDB: Initializing buffer pool, total size = 128.000MiB, chunk size = 2.000MiB
2024-07-16 21:16:42 0 [Note] InnoDB: Completed initialization of buffer pool
2024-07-16 21:16:42 0 [Note] InnoDB: End of log at LSN=47749
2024-07-16 21:16:42 0 [Note] InnoDB: Opened 3 undo tablespaces
2024-07-16 21:16:42 0 [Note] InnoDB: 128 rollback segments in 3 undo tablespaces are active.
2024-07-16 21:16:42 0 [Note] InnoDB: Setting file './ibtmp1' size to 12.000MiB. Physically writing the file full; Please wait ...
2024-07-16 21:16:42 0 [Note] InnoDB: File './ibtmp1' size is now 12.000MiB.
2024-07-16 21:16:42 0 [Note] InnoDB: log sequence number 47749; transaction id 15
2024-07-16 21:16:42 0 [Note] Plugin 'FEEDBACK' is disabled.
2024-07-16 21:16:42 0 [Note] InnoDB: Loading buffer pool(s) from /opt/homebrew/var/mysql/ib_buffer_pool
2024-07-16 21:16:42 0 [Note] Plugin 'wsrep-provider' is disabled.
2024-07-16 21:16:42 0 [Note] InnoDB: Buffer pool(s) load completed at 240716 21:16:42
2024-07-16 21:16:42 0 [Note] Server socket created on IP: '127.0.0.1'.
2024-07-16 21:16:42 0 [ERROR] Can't start server : Bind on unix socket: Address already in use
2024-07-16 21:16:42 0 [ERROR] Do you already have another server running on socket: /tmp/mysql.sock ?
2024-07-16 21:16:42 0 [ERROR] Aborting
240716 21:16:43 mysqld_safe mysqld from pid file /opt/homebrew/var/mysql/capel0084052675..pid ended
```

Any suggestions how do I proceed, attaching my.cnf below?

➜ cat /opt/homebrew/etc/my.cnf
```sh
# my.cnf

#
# This group is read both by the client and the server
# use it for options that affect everything
#
[client-server]

#
# include *.cnf from the config directory
#
!includedir /opt/homebrew/etc/my.cnf.d

[mysqld]
character-set-client-handshake = FALSE
character-set-server = utf8mb4
collation-server = utf8mb4_unicode_ci
bind-address = 127.0.0.1
socket = /tmp/mysql.sock

[mysql]
default-character-set = utf8mb4
```

**Reply by gromgit** [reply]
> ```
> 2024-07-16 21:16:42 0 [ERROR] Can't start server : Bind on unix socket: Address already in use
> 2024-07-16 21:16:42 0 [ERROR] Do you already have another server running on socket: /tmp/mysql.sock ?
> 2024-07-16 21:16:42 0 [ERROR] Aborting
> ```

Do you have another mysql process running? What's the output of `pgrep -fl mysql mariadb`?

**Reply by maazmmd** [reply]
I don't think so its running, I don't see on activity monitor on mac but i tried on terminal but was not able to figure it out
```sh
➜ pgrep -fl mysql mariadb     
13896 /Applications/Visual Studio Code.app/Contents/MacOS/Electron /opt/homebrew/opt/mariadb/.brew/mariadb.rb

➜ ps aux | grep mysql
mohmaaz          49152   0.0  0.0 410063264    160 s005  R+   10:59am   0:00.00 grep --color=auto --exclude-dir=.bzr --exclude-dir=CVS --exclude-dir=.git --exclude-dir=.hg --exclude-dir=.svn --exclude-dir=.idea --exclude-dir=.tox mysql

➜ ps aux | grep mariadb
mohmaaz          13896   0.0  0.6 1596626672 207488   ??  S     8:46pm   0:06.34 /Applications/Visual Studio Code.app/Contents/MacOS/Electron /opt/homebrew/opt/mariadb/.brew/mariadb.rb
mohmaaz          49292   0.0  0.0 410063264    192 s005  R+   10:59am   0:00.00 grep --color=auto --exclude-dir=.bzr --exclude-dir=CVS --exclude-dir=.git --exclude-dir=.hg --exclude-dir=.svn --exclude-dir=.idea --exclude-dir=.tox mariadb
```

**Reply by gromgit** [reply]
What does `lsof /tmp/mysql.sock` output?

**Reply by maazmmd** [reply]
```sh
➜ lsof /tmp/mysql.sock
COMMAND    PID    USER   FD   TYPE             DEVICE SIZE/OFF NODE NAME
mariadbd 66984 mohmaaz   25u  unix 0xedbfea7a63d0ad4e      0t0      /tmp/mysql.sock
```

**Reply by gromgit** [reply]
There's your stray `mariadbd` process. `ps -ef | grep 66984` to see what exactly was run, that might also give a clue as to why it didn't appear in any of the `ps/pgrep` outputs.

**Reply by maazmmd** [reply]
I restarted my mac, as I was getting another issue with port already in use. Here's the output of stray `mariadbd` process

```sh
➜ ps -ef | grep 1341
1013988708  1341  1110   0 11:59am ??         0:00.83 /opt/homebrew/opt/mariadb/bin/mariadbd --basedir=/opt/homebrew/opt/mariadb --datadir=/opt/homebrew/var/mysql --plugin-dir=/opt/homebrew/opt/mariadb/lib/plugin --log-error=/opt/homebrew/var/mysql/capel0084052675..err --pid-file=capel0084052675..pid --socket=/tmp/mysql.sock
1013988708  2358  2357   0 12:01pm ttys003    0:00.01 q _ inline-shell-completion --buffer ps -ef | grep 1341
1013988708  2361  1769   0 12:01pm ttys003    0:00.00 grep --color=auto --exclude-dir=.bzr --exclude-dir=CVS --exclude-dir=.git --exclude-dir=.hg --exclude-dir=.svn --exclude-dir=.idea --exclude-dir=.tox 1341
```

**Reply by maazmmd** [reply]
Fixed the issue ➜ gave permissions to /tmp and .err and .pid files as mariadb was unable to read/write files and adding `tmpdir = /tmp` in `my.cnf`. Thanks for your help @gromgit

```sh
# add tmp directory in my.cnf
tmpdir          = /tmp

# Grant permissions
sudo chmod 0777 /tmp
sudo chmod 0777 /opt/homebrew/var/mysql/{your_mac_profile}..err
sudo chmod 0777 /opt/homebrew/var/mysql/{your_mac_profile}..pid
chmod 0777 /opt/homebrew/var/{your_mac_profile}..err
```

Reference: https://stackoverflow.com/questions/8633373/mysql-cant-create-write-to-file-error13

**Comment by draven3166** [comment]
There are many cases for Mariadb or any other service to stop working properly on the computer. _My case MacOs Sequoia_
causes
- Sometimes a memory error and/or ram crash. 
- OS error 
  
They can be the specific cause in this type of problems and for that a generic question in the communities to it does not always lead you to the **answer/solution** valid since the causes are several/different.
In my case I have followed the guidelines above but it did not solve the problem. And one of those was to execute the process of direct call to mariadb and its handler that uses the socket by default, this when executed tells you which threads and dependencies it has, it draws you technically where everything happens, besides showing you how it works.
`# mysqld`
**preview it is important to see** 
`# mysqld --verbose --help | grep “socket”   `
To locate exactly where your configurations are in the system and what values they are taking.
**_My specific case was a memory error and a break when initializing services and dependencies which broke everything and had no service and could not schedule a three hours!!!_**

It is solved by making a backup of the **tc.log** file (same path in #[coment](#solution-for-mysql--mariadb-start-issue-on-macos)) delete the last one, this case a read and follow [this thread ](https://github.com/orgs/Homebrew/discussions/3410#discussioncomment-10069601) and _not have the file mysql.sock_ in place. And run.
`# mysqld` 
so that the instances in **mysql.servers** regenerate the paths and processes to regenerate the path and be able to start the service with the correct configuration without breaking or forcing any process to act arbitrarily because that is the workflow where there is a problem you have to investigate to solve it and start everything again well.
Do yourself a favor as a developer read the warnings that the prompt returns and the errors that cause them, it will help you a lot.
>[!NOTE]
>Dont forget restart service to include last configuration in your pc. 
Happy code.
